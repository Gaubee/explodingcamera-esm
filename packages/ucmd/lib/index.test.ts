import { describe, expect, test } from "vitest";
import { createCommand, ucmd } from ".";
import type { CommandContext, Command } from ".";

describe("parse commands", () => {
	test("ucmd", () => {
		let x = ucmd("example")
			.withBaseCommand({
				args: {
					foo: true,
					bar: false,
				},
			})
			.withCommand({
				name: "build",
				args: {
					test: {
						type: "string",
						default: "test",
					},
					foo: true,
					bar: false,
				},
			});

		expect(x.parse(["build", "test1", "--foo", "--test", "test", "test2"])).toMatchInlineSnapshot(`
			{
			  "res": {
			    "positionals": [
			      "test1",
			      "test2",
			    ],
			    "values": {
			      "foo": true,
			      "test": "test",
			    },
			  },
			  "run": [Function],
			}
		`);

		expect(x.parse(["test1", "--foo", "test", "test2"])).toMatchInlineSnapshot(`
			{
			  "res": {
			    "positionals": [
			      "test1",
			      "test",
			      "test2",
			    ],
			    "values": {
			      "foo": true,
			    },
			  },
			  "run": [Function],
			}
		`);
	});
});

describe("basic api", () => {
	test("ucmd", () => {
		let x = ucmd("example");
		expect(x).toBeDefined();
		expect(x.state.name).toBe("example");
	});

	test("withName", () => {
		let x = ucmd("example").withName("overwritten");
		expect(x.state.name).toBe("overwritten");
	});

	test("withCommand", () => {
		let res = ucmd("example").withCommand({
			name: "build",
			args: {
				foo: true,
				bar: false,
			},
		});

		expect(res.state.commands.build).toBeDefined();
	});

	test("withImportedCommand", () => {
		let one = ucmd("example").withCommand({
			name: "build",
			args: {
				foo: true,
				bar: false,
			},
		});

		let twoCmd = createCommand({
			name: "build",
			args: {
				foo: true,
				bar: {},
				baz: false,
			},
		});

		let twoCmdDirect = {
			name: "build",
			args: {
				foo: true,
				bar: {},
				baz: false,
			},
		} satisfies Command;

		expect(twoCmd).toEqual(twoCmdDirect);

		let twoFn = (ctx: CommandContext<typeof twoCmd>) => {
			ctx.args.foo;
		};

		let two = ucmd("example").withCommand(twoCmd, twoFn);

		expect(one).toEqual(two);
	});
});
