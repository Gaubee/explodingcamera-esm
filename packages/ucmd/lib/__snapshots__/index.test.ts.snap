// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`parse commands ucmd 1`] = `
{
  "res": {
    "positionals": [
      "test1",
      "test2",
    ],
    "values": {
      "foo": true,
      "test": "test",
    },
  },
  "run": [Function],
}
`;

exports[`parse commands ucmd 2`] = `
{
  "res": {
    "positionals": [
      "test1",
      "test",
      "test2",
    ],
    "values": {
      "foo": true,
    },
  },
  "run": [Function],
}
`;
